core/src/main/scala/akka/kafka/internal/SubSourceLogic.scala

```diff
@@ -273,98 +290,160 @@ private abstract class SubSourceLogic[K, V, Msg](
     tp: TopicPartition,
     consumerActor: ActorRef,
-    subSourceStartedCb: AsyncCallback[(TopicPartition, Control)],
-    subSourceCancelledCb: AsyncCallback[(TopicPartition, Option[ConsumerRecord[K, V]])],
-    messageBuilder: MessageBuilder[K, V, Msg],
-    actorNumber: Int
+    subSourceStartedCb: AsyncCallback[(TopicPartition, ControlAndStageActor)],
+    subSourceCancelledCb: AsyncCallback[(TopicPartition, SubSourceCancellationStrategy)],
+    actorNumber: Int,
+    subSourceStageLogicFactory: SubSourceStageLogicFactory[K, V, Msg]
 ) extends GraphStage[SourceShape[Msg]] { stage =>
+
   val out = Outlet[Msg]("out")
-  val shape = new SourceShape(out)
+  val shape: SourceShape[Msg] = new SourceShape(out)
 
   override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
-    new GraphStageLogic(shape) with PromiseControl with MetricsControl with StageLogging {
-      override def executionContext: ExecutionContext = materializer.executionContext
-      override def consumerFuture: Future[ActorRef] = Future.successful(consumerActor)
-      val shape = stage.shape
-      val requestMessages = KafkaConsumerActor.Internal.RequestMessages(0, Set(tp))
-      var requested = false
-      var subSourceActor: StageActor = _
-      var buffer: Iterator[ConsumerRecord[K, V]] = Iterator.empty
-
-      override def preStart(): Unit = {
-        log.debug("#{} Starting SubSource for partition {}", actorNumber, tp)
-        super.preStart()
-        subSourceStartedCb.invoke(tp -> this.asInstanceOf[Control])
-        subSourceActor = getStageActor {
-          case (_, msg: KafkaConsumerActor.Internal.Messages[K, V]) =>
-            requested = false
-            // do not use simple ++ because of https://issues.scala-lang.org/browse/SI-9766
-            if (buffer.hasNext) {
-              buffer = buffer ++ msg.messages
-            } else {
-              buffer = msg.messages
-            }
-            pump()
-          case (_, Status.Failure(e)) =>
-            failStage(e)
-          case (_, Terminated(ref)) if ref == consumerActor =>
-            failStage(new ConsumerFailed)
-        }
-        subSourceActor.watch(consumerActor)
-        consumerActor.tell(RegisterSubStage, subSourceActor.ref)
-      }
+    subSourceStageLogicFactory.create(shape, tp, consumerActor, subSourceStartedCb, subSourceCancelledCb, actorNumber)
+}
 
-      override def postStop(): Unit = {
-        onShutdown()
-        super.postStop()
+/** Internal API
+ *
+ * SubSourceStageLogic [[akka.kafka.scaladsl.Consumer.Control]] and the stage actor [[ActorRef]]
+ */
+@InternalApi
+private final case class ControlAndStageActor(control: Control, stageActor: ActorRef)
+
+sealed trait SubSourceCancellationStrategy
+final case class SeekToOffsetAndReEmit(offset: Long) extends SubSourceCancellationStrategy
+case object ReEmit extends SubSourceCancellationStrategy
+case object DoNothing extends SubSourceCancellationStrategy
```

Comment:
Explain this concept.

Reply:
This decides if the sub source is re-emmitted downstream (after possibly re-seeking to last committed offsets).  For the transactional partitioned source we want to stop, fail, or shutdown the outer `SubSourceStage` if any sub source cancels.  I didn't change the default behaviour which is now implemented as a combination of `SeekToOffsetAndReEmit` and `ReEmit`.  I'll add comments.